src/util/dependent-resource.ts

```diff
@@ -20,58 +20,70 @@ key-value resources directly attached to a parent (e.g. tags, config variables).
 */
 
 import { isId, isUnauthorizedResponse, mergePineOptions } from '../util';
-import type { Pine, PineOptions } from '..';
-import type { Dictionary } from '../../typings/utils';
+import type { BalenaModel, PineClient } from '..';
+import type { Dictionary, StringKeyof } from '../../typings/utils';
+import type {
+	ExpandableStringKeyOf,
+	FilterObj,
+	ODataOptionsWithoutCount,
+	UpsertParams,
+} from 'pinejs-client-core';
 
-interface DependentResource {
-	id: number;
-	value: string;
-}
+type DependentResource = {
+	[K in StringKeyof<BalenaModel>]: 'Read' extends keyof BalenaModel[K]
+		? BalenaModel[K]['Read'] extends { id: number; value: string }
+			? BalenaModel[K]['Write'] extends { value: string }
+				? K
+				: never
+			: never
+		: never;
+}[StringKeyof<BalenaModel>];
```

Comment:
I guess this can be simplified to a single structural check. Not sure whether it's faster, but it looks easier to read imo
```suggestion
type DependentResource = {
	[K in StringKeyof<BalenaModel>]: BalenaModel[K] extends {
		'Read': {
			id: number;
			value: string;
		},
		'Write': { value: string; }
	}
		? K
		: never
}[StringKeyof<BalenaModel>];
```